p2p/peer.go

```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
 	m.mtx.Lock()
 	defer m.mtx.Unlock()
 
+	delete(m.dialing, peerID)
+
+	if m.connected[peerID] {
+		return fmt.Errorf("peer %v is already connected", peerID)
+	}
+	if m.options.MaxConnected > 0 &&
+		len(m.connected) >= int(m.options.MaxConnected)+int(m.options.MaxConnectedUpgrade) {
+		return fmt.Errorf("already connected to maximum number of peers")
+	}
+
 	peer, err := m.store.Get(peerID)
 	if err != nil {
 		return err
 	} else if peer == nil {
-		return fmt.Errorf("unknown peer %q", peerID)
+		return fmt.Errorf("peer %q was removed while dialing", peerID)
 	}
+	m.connected[peerID] = true
 
-	if m.connected[peerID] {
-		return fmt.Errorf("peer %v is already connected", peerID)
+	now := time.Now().UTC()
+	peer.LastConnected = now
+	if addressInfo := peer.LookupAddressInfo(address); addressInfo != nil {
+		addressInfo.DialFailures = 0
+		addressInfo.LastDialSuccess = now
 	}
-	delete(m.dialing, peerID)
-	m.connected[peerID] = true
-	return nil
+	return m.store.Set(peer)
 }
 
 // Accepted marks an incoming peer connection successfully accepted. If the peer
-// is already connected this will return an error.
+// is already connected or we don't allow additional connections then this will
+// return an error.
+//
+// If MaxConnectedUpgrade is non-zero, the accepted peer is better-scored than any
+// other connected peer, and the number of connections does not exceed
+// MaxConnected + MaxConnectedUpgrade then we accept the connection and rely on
+// EvictNext() to evict lower-scored peers.
 //
 // NOTE: We can't take an address here, since e.g. TCP uses a different port
 // number for outbound traffic than inbound traffic, so the peer's endpoint
 // wouldn't necessarily be an appropriate address to dial.
-func (m *peerManager) Accepted(peerID NodeID) error {
+func (m *PeerManager) Accepted(peerID NodeID) error {
 	m.mtx.Lock()
 	defer m.mtx.Unlock()
+
+	if m.connected[peerID] {
+		return fmt.Errorf("peer %q is already connected", peerID)
+	}
+	if m.options.MaxConnected > 0 &&
+		len(m.connected) >= int(m.options.MaxConnected)+int(m.options.MaxConnectedUpgrade) {
+		return fmt.Errorf("already connected to maximum number of peers")
+	}
+
 	peer, err := m.store.Get(peerID)
 	if err != nil {
 		return err
-	} else if peer == nil {
-		peer = newPeerInfo(peerID)
-		if err = m.store.Set(peer); err != nil {
-			return err
+	}
+	if peer == nil {
+		peer = &peerInfo{
+			ID:         peerID,
+			Persistent: m.options.isPersistent(peerID),
 		}
 	}
-	if m.connected[peerID] {
-		return fmt.Errorf("peer %q is already connected", peerID)
+
+	// If we're already full (i.e. at MaxConnected), but we allow upgrades (and we
+	// know from the check above that we have upgrade capacity), then we can look
+	// for a lower-scored evictable peer, and if found we can accept this connection
+	// anyway and let EvictNext() evict the lower-scored peer for us.
+	//
+	// FIXME: There is a race condition here where, if there exists a single
```

Comment:
How do you propose we go about fixing this? If you haven't already, let's be sure to mark this in #5670.

Reply:
Probably by keeping an `upgrading` map that reserves an eviction slot for a higher peer.

Reply:
Tried implementing this, but the logic became a bit more complicated than I'd like. Will revisit this when I attempt the channel-based API, since that might reorganize this logic a bit anyway (i.e. by pushing the peer to the eviction queue immediately instead of relying on `EvictNext` to determine it later).